selfdrive/loggerd/v4l_encoder.cc

```diff
@@ -0,0 +1,380 @@
+  }
+
+  // VideoWriter is freed on out of scope
+}
+
+void V4LEncoder::dequeue_handler(V4LEncoder *e) {
+  e->segment_num++;
+  uint32_t idx = 0;
+  bool exit = false;
+
+  // POLLIN is capture, POLLOUT is frame
+  struct pollfd pfd;
+  pfd.events = POLLIN | POLLOUT;
+  pfd.fd = e->fd;
+
+  while (!exit) {
+    int rc = poll(&pfd, 1, 1000);
+    if (!rc) { printf("poll timeout"); continue; }
+
+    /*struct timespec t;
+    clock_gettime(CLOCK_BOOTTIME, &t);
+    uint64_t current_time = t.tv_sec * 1000000000ULL + t.tv_nsec;
+    printf("poll %x at %f ms\n", pfd.revents, current_time/1e6);*/
+
+    if (pfd.revents & POLLIN) {
+      unsigned int bytesused, flags, index;
+      struct timeval timestamp;
+      int ret = dequeue_buffer(e->fd, V4L2_BUF_TYPE_VIDEO_CAPTURE_MPLANE, &index, &bytesused, &flags, &timestamp);
+      assert(ret == 0);
+      e->buf_out[index].sync(VISIONBUF_SYNC_FROM_DEVICE);
+      uint8_t *buf = (uint8_t*)e->buf_out[index].addr;
+
+      // eof packet, we exit
+      if (flags & V4L2_QCOM_BUF_FLAG_EOS) exit = true;
+      uint64_t ts = timestamp.tv_sec * 1000000 + timestamp.tv_usec;
+
+      // broadcast packet
+      MessageBuilder msg;
+      auto event = msg.initEvent(true);
+      auto edata = (e->type == DriverCam) ? event.initDriverEncodeData() :
+        ((e->type == WideRoadCam) ? event.initWideRoadEncodeData() :
+        (e->remuxing ? event.initQRoadEncodeData() : event.initRoadEncodeData()));
+      edata.setData(kj::arrayPtr<capnp::byte>(buf, bytesused));
+      edata.setTimestampEof(ts);
+      edata.setIdx(idx++);
+      edata.setSegmentNum(e->segment_num);
+      edata.setFlags(flags);
+
+      auto words = new kj::Array<capnp::word>(capnp::messageToFlatArray(msg));
+      auto bytes = words->asBytes();
+      e->pm->send(e->service_name, bytes.begin(), bytes.size());
+      if (e->write) {
+        e->to_write.push(words);
+      } else {
+        delete words;
+      }
+
+      if (env_debug_encoder) {
+        printf("%20s got %6d bytes in buffer %d with flags %8x and ts %lu lat %.2f ms (%lu frames free)\n",
+          e->filename, bytesused, index, flags, ts, ((event.getLogMonoTime() / 1000)-ts)/1000., e->free_buf_in.size());
+      }
+
+      // requeue the buffer
+      ret = queue_buffer(e->fd, V4L2_BUF_TYPE_VIDEO_CAPTURE_MPLANE, index, &e->buf_out[index]);
+      assert(ret == 0);
+    }
+
+    if (pfd.revents & POLLOUT) {
+      unsigned int index;
+      int ret = dequeue_buffer(e->fd, V4L2_BUF_TYPE_VIDEO_OUTPUT_MPLANE, &index);
+      assert(ret == 0);
+      e->free_buf_in.push(index);
+    }
+  }
+}
+
+V4LEncoder::V4LEncoder(
+  const char* filename, CameraType type, int in_width, int in_height,
+  int fps, int bitrate, bool h265, int out_width, int out_height, bool write)
+  : type(type), in_width_(in_width), in_height_(in_height),
+    width(out_width), height(out_height), fps(fps),
+    filename(filename), remuxing(!h265), write(write) {
+  fd = open("/dev/video33", O_RDWR|O_NONBLOCK);
```

Comment:
Can we open this by name like we do for cameras?

Reply:
Done